### PR TITLE
#1352: Ensure lqip data is always available.

### LINKED ISSUE #1352: Unable to get LQIP data
## Description

When attempting to fetch the `lqip_data` field from the database via the `HasMedia->lowQualityImagePlaceholder()` method the transparent fallback is always returned because `$media->pivot` contains everything you'd expect except `lqip_data`

### Steps to reproduce

1. Upload an image
2. Attach the image to something - aka populate the mediables table
3. Generate the LQIP data (`php artisan twill:lqip`)
4. Attempt to read the `lqip_data` via the `HasMedia->lowQualityImagePlaceholder()` method

### Expected result

The value of `lqip_data` is returned from the DB

### Actual result

The `ImageService::getTransparentFallbackUrl()` is retuned

### Versions

Twill version: 2.6.0
Laravel version: 8.78.1
PHP version: 8.1.0
Database engine: MySQL 5.7.34

### PR DESCRIPTION
**Pull request by haringsrob**
## Description

This pr ensures that lqip_data is always passed around to make sure the get method works.

If there are more ways to get lqip data let me know so I can double check.

I also made a change in the method for generating lqip, but please have a look at it yourself @ifox  as I am not entirely sure of that solution.

## Related Issues

Fixes #1352

### CODE REVIEW

File: src/Commands/RefreshLQIP.php [hunk 1]
```diff
@@ -66,14 +67,24 @@ public function handle()
...
+
                     $url = ImageService::getLQIPUrl($uuid, $crop_params + ['w' => $lqip_width]);
 
+                    if ($imageService === Glide::class) {
+                        // Glide is a local imaging service so we should make sure we also utilize the local url.
+                        // This solution might not work at all in some cases.
+                        // I feel like this is not optimal and we could maybe directly use the glide api in this case
+                        // but maybe that is not worth the effort (for now).
+                        $url = url($url);
+                    }
```
Comment (freshleafmedia) [inline src/Commands/RefreshLQIP.php:80]: Not sure this is necessary. While trying to get Glide working locally I found I just needed to set the following env variables locally (Worth noting I also had previously run `php artisan storage:link`):

```
GLIDE_BASE_URL=http://www.localhost
GLIDE_BASE_PATH=/storage
```

Perhaps this just needs documenting?

File: src/Commands/RefreshLQIP.php [hunk 2]
```diff
@@ -66,14 +67,24 @@ public function handle()
...
+
                     try {
                         $data = file_get_contents($url);
                         $dataUri = 'data:image/gif;base64,' . base64_encode($data);
                         $this->db->table(config('twill.mediables_table', 'twill_mediables'))->where('id', $attached_media->id)->update(['lqip_data' => $dataUri]);
                     } catch (\Exception $e) {
-                        $this->info("LQIP was not generated for " . $uuid);
+                        $this->info("LQIP was not generated for $uuid because {$e->getMessage()}");
```
Comment (freshleafmedia) [inline src/Commands/RefreshLQIP.php:87]: :+1: I was going to PR something similar to this. Catching `\Exception` with no user feedback is too vague

File: src/Services/MediaLibrary/Glide.php [hunk 3]
```diff
@@ -286,9 +286,6 @@ private function getOriginalMediaUrl($id)
             return null;
         }
 
-        /** @var string $endpoint */
-        $endpoint;
```
Comment (freshleafmedia) [inline src/Services/MediaLibrary/Glide.php:290]: By removing this wont we get an Undefined Variable error if `$endpointType` is not `local`, `s3` or `azure`?

File: src/Commands/RefreshLQIP.php (hunk 1 above)
Comment (freshleafmedia) [inline src/Commands/RefreshLQIP.php:80]: Not sure this is necessary. While trying to get Glide working locally I found I just needed to set the following env variables locally (Worth noting I also had previously run `php artisan storage:link`):

```
GLIDE_BASE_URL=http://www.localhost
GLIDE_BASE_PATH=/storage
```

Perhaps this just needs documenting?

File: src/Services/MediaLibrary/Glide.php (hunk 3 above)
Comment (haringsrob) [reply src/Services/MediaLibrary/Glide.php:290]: You are right!

File: src/Commands/RefreshLQIP.php (hunk 1 above)
Comment (haringsrob) [reply src/Commands/RefreshLQIP.php]: Thanks @freshleafmedia.

I did not know. But I changed it to show a message now if it is missing: 

```
Cannot generate LQIP. Missing glide base url. Please set GLIDE_BASE_URL in your .env
```

Comment (haringsrob) [reply src/Commands/RefreshLQIP.php]: @ifox maybe we can make twill.glide.base_url default to the app.url?

File: src/Repositories/Behaviors/HandleMedias.php [hunk 4]
```diff
@@ -91,6 +91,7 @@ private function getMedias($fields)
                                     'crop_h' => $cropData['height'],
                                     'crop_x' => $cropData['x'],
                                     'crop_y' => $cropData['y'],
+                                    'lqip_data' => $media['lqip_data'] ?? null,
```
Comment (ifox) [inline src/Repositories/Behaviors/HandleMedias.php:94]: See my other comment, I'm not sure we need to add it here and below.

**Review by ifox — CHANGES_REQUESTED** [review]

File: src/Models/Media.php [hunk 5]
```diff
@@ -78,6 +77,7 @@ public function toCmsArray()
             'tags' => $this->tags->map(function ($tag) {
                 return $tag->name;
             }),
+            'lqip_data' => $this->pivot->lqip_data ?? null,
```
[the remark this answers is not in the document]
Comment (haringsrob) [reply src/Models/Media.php]: So I removed those and still seems to work all fine.

**Review by ifox — APPROVED** [review]

**Pull request merged**